src/matlab2tikz.m

```diff
@@ -4817,22 +4830,50 @@ case size(cData,2) % specific y-data is given
             position(i,:) = axesPos ./ [figureSize, figureSize];
 
         end
-
-        % Change size if DataAspectRatioMode is manual
-        if isequal(lower(get(axesHandle,'DataAspectRatioMode')),'manual')
-            % get limits
-            xLim = get(axesHandle, 'XLim');
-            yLim = get(axesHandle, 'YLim');
-            % Get Aspect Ratio between width and height
-            aspectRatio = get(axesHandle,'DataAspectRatio');
-            % And Adjust it to the figure dimensions
-            aspectRatio = aspectRatio(1) * figWidth * (yLim(2) - yLim(1)) ...
-                / (aspectRatio(2) * figHeight * (xLim(2)-xLim(1)));
-            % Recompute height
-            newHeight = position(i,3) * aspectRatio;
-            % shrink width if newHeight is too large
-            if newHeight > position(i,4)
-                % Recompute width
+        
+        if ~(strcmpi(get(axesHandle, 'DataAspectRatioMode'), 'auto') ...
+                && strcmpi(get(axesHandle, 'PlotBoxAspectRatioMode'), 'auto'))
+            limits = axis(axesHandle);
+            if ~strcmpi(get(axesHandle, 'DataAspectRatioMode'), 'manual')
+                limits = mod(0:length(limits)-1,2); % plot box limitting mode
+                aspectRatio = get(axesHandle, 'PlotBoxAspectRatio');
+            else
+                aspectRatio = get(axesHandle, 'DataAspectRatio');
+            end
+            
+            if length(limits) == 4
+                % 2d view
+                aspectRatio = aspectRatio(1) * figWidth * (limits(4) - limits(3)) ...
+                    / (aspectRatio(2) * figHeight * (limits(2) - limits(1)));
+            else
+                % 3d view
+                
+                % project vertices of 3d plot box (this results in 2d coordinates in
+                % an absolute coordinate system that is scaled proportionally by
+                % Matlab to fit the axes position box)
+                projection = view(axesHandle);
+                vertex1 = projection * [0;0;0;1];
+                vertex2 = projection * [1;0;0;1];
+                vertex3 = projection * [0;1;0;1];
+                vertex4 = projection * [0;0;1;1];
+                vertex5 = projection * [1;1;0;1];
+                vertex6 = projection * [1;0;1;1];
+                vertex7 = projection * [0;1;1;1];
+                vertex8 = projection * [1;1;1;1];
```

Comment:
I double checked this and it is indeed equivalent. Also note that this calculation works equally well for 2d views so we could simplify the code even further as 2d and 3d views are equivalent in Matlab. 

Note that this calculation does not work correctly for perspective projections (the plot box is no longer necessarily constrained by the long diagonals for those). Pgfplots does not support perspective projections as far as I know, so maybe we should display a warning in that case?